File: cpp/src/protocol/protocol.cpp
#include "dlt645/protocol/protocol.h"
#include <algorithm>
#include <stdexcept>
#include <array>

namespace dlt645
{
    namespace protocol
    {

        // 解码数据域（±33H转换）
        std::vector<uint8_t> Frame::decodeData(std::span<const uint8_t> data)
        {
            std::vector<uint8_t> result(data.size());
            std::ranges::transform(data, result.begin(), [](uint8_t b)
                                   { return b - 0x33; });
            return result;
        }

        // 编码数据域（±33H转换）
        std::vector<uint8_t> Frame::encodeData(std::span<const uint8_t> data)
        {
            std::vector<uint8_t> result(data.size());
            std::ranges::transform(data, result.begin(), [](uint8_t b)
                                   { return b + 0x33; });
            return result;
        }

        // 计算校验和（模256求和）
        uint8_t Frame::calculateChecksum(std::span<const uint8_t> data)
        {
            uint8_t sum = 0;
            for (const auto &b : data)
            {
                sum += b;
            }
            return sum;
        }

        // 构建帧
        std::vector<uint8_t>
        Frame::buildFrame(std::span<const uint8_t, 6> addr, uint8_t ctrlCode, const std::vector<uint8_t> &data)
        {
            std::vector<uint8_t> buf;
            buf.reserve(12 + data.size()); // 预留空间

            buf.push_back(FRAME_START_BYTE);
            buf.insert(buf.end(), addr.begin(), addr.end());
            buf.push_back(FRAME_START_BYTE);
            buf.push_back(ctrlCode);

            // 数据域编码
            auto encodedData = encodeData(data);
            buf.push_back(static_cast<uint8_t>(encodedData.size()));
            buf.insert(buf.end(), encodedData.begin(), encodedData.end());

            // 计算校验和
            uint8_t checkSum = calculateChecksum(buf);
            buf.push_back(checkSum);
            buf.push_back(FRAME_END_BYTE);

            // 前导字节添加
            std::array<uint8_t, 4> preamble = {0xFE, 0xFE, 0xFE, 0xFE};
            buf.insert(buf.begin(), preamble.begin(), preamble.end());
            return buf;
        }

        // 序列化帧
        std::vector<uint8_t> Frame::serialize() const
        {
            if (startFlag != FRAME_START_BYTE || endFlag != FRAME_END_BYTE)
            {
                throw std::runtime_error("Invalid start or end flag");
            }

            std::vector<uint8_t> buf;
            buf.reserve(preamble.size() + 12 + data.size()); // 预留空间

            // 写入前导字节
            buf.insert(buf.end(), preamble.begin(), preamble.end());

            // 写入起始符
            buf.push_back(startFlag);

            // 写入地址
            buf.insert(buf.end(), addr.begin(), addr.end());

            // 写入第二个起始符
            buf.push_back(startFlag);

            // 写入控制码
            buf.push_back(ctrlCode);

            // 数据域编码
            auto encodedData = encodeData(data);

            // 写入数据长度
            buf.push_back(static_cast<uint8_t>(encodedData.size()));

            // 写入编码后的数据
            buf.insert(buf.end(), encodedData.begin(), encodedData.end());

            // 计算并写入校验和
            uint8_t checkSum = calculateChecksum(buf);
            buf.push_back(checkSum);

            // 写入结束符
            buf.push_back(endFlag);

            return buf;
        }

        // 反序列化帧
        std::shared_ptr<Frame> Frame::deserialize(const std::vector<uint8_t> &raw)
        {
            // 基础校验
            if (raw.size() < 12)
            { // 最小帧长度=起始符(1)+地址(6)+起始符(1)+控制码(1)+数据长度(1)+结束符(1)
                return nullptr;
            }

            // 帧边界检查
            auto startIt = std::ranges::find(raw, FRAME_START_BYTE);
            if (startIt == raw.end() || std::distance(startIt, raw.end()) < 11)
            {
                return nullptr;
            }

            size_t startIdx = std::distance(raw.begin(), startIt);
            if (startIdx + 7 >= raw.size() || raw[startIdx + 7] != FRAME_START_BYTE)
            { // 第二个起始符位置校验
                return nullptr;
            }

            // 构建帧结构
            auto frame = std::make_shared<Frame>();
            frame->startFlag = raw[startIdx];
            std::copy_n(raw.begin() + startIdx + 1, 6, frame->addr.begin());
            frame->ctrlCode = raw[startIdx + 8];
            frame->dataLen = raw[startIdx + 9];

            // 数据域提取
            size_t dataStart = startIdx + 10;
            size_t dataEnd = dataStart + frame->dataLen;
            if (dataEnd > raw.size() - 2)
            { // 预留校验位和结束符
                return nullptr;
            }

            // 数据域解码
            frame->data = decodeData({raw.begin() + dataStart, frame->dataLen});

            // 校验和验证（从第一个68H到校验码前）
            std::vector<uint8_t> checkData(raw.begin() + startIdx, raw.begin() + dataEnd);
            uint8_t calculatedSum = calculateChecksum(checkData);

            if (calculatedSum != raw[dataEnd])
            {
                return nullptr;
            }

            // 结束符验证
            if (dataEnd + 1 >= raw.size() || raw[dataEnd + 1] != FRAME_END_BYTE)
            {
                return nullptr;
            }

            frame->checkSum = raw[dataEnd];
            frame->endFlag = raw[dataEnd + 1];

            return frame;
        }

    } // namespace protocol
} // namespace dlt645

File: python/src/protocol/protocol.py
"""DL/T 645-2007 数据帧编码、校验与流式解析。"""

from typing import Optional

from .frame import FRAME_END_BYTE, FRAME_START_BYTE, Frame
from .log import log


class DLT645Protocol:
    """无状态的 DL/T 645 帧编解码器。"""

    MAX_DATA_LENGTH = 0xFF
    MAX_PREAMBLE_LENGTH = 4
    MIN_FRAME_LENGTH = 12

    @classmethod
    def decode_data(cls, data: bytes) -> bytes:
        """移除数据域逐字节增加的 ``0x33`` 偏移。"""
        return bytes((value - 0x33) & 0xFF for value in data)

    @classmethod
    def encode_data(cls, data: bytes) -> bytes:
        """按协议为数据域逐字节增加 ``0x33`` 偏移。"""
        return bytes((value + 0x33) & 0xFF for value in data)

    @classmethod
    def calculate_checksum(cls, data: bytes) -> int:
        """计算从第一个 ``0x68`` 到数据域末尾的模 256 累加和。"""
        return sum(data) & 0xFF

    @classmethod
    def build_frame(
        cls,
        addr: bytes,
        ctrl_code: int,
        data: Optional[bytes],
        preamble_count: int = 4,
    ) -> bytearray:
        """从解码后的地址、控制码和数据域构建完整帧。"""
        data = data or b""
        if len(addr) != 6:
            raise ValueError("地址长度必须为6字节")
        if not 0 <= ctrl_code <= 0xFF:
            raise ValueError("控制码必须在0~255范围内")
        if len(data) > cls.MAX_DATA_LENGTH:
            raise ValueError("数据域长度不能超过255字节")
        if not 0 <= preamble_count <= cls.MAX_PREAMBLE_LENGTH:
            raise ValueError("前导字节数量必须在0~4范围内")

        frame = Frame(
            preamble=bytes([0xFE] * preamble_count),
            start_flag=FRAME_START_BYTE,
            addr=addr,
            ctrl_code=ctrl_code,
            data_len=len(data),
            data=data,
            end_flag=FRAME_END_BYTE,
        )
        return bytearray(cls.serialize(frame))

    @classmethod
    def deserialize(cls, raw: bytes) -> Optional[Frame]:
        """解析第一帧；数据不完整或没有合法帧时抛出 ``ValueError``。"""
        _, frame = cls.deserialize_with_remaining(raw)
        if frame is None:
            raise ValueError("No complete frame found")
        return frame

    @staticmethod
    def _trailing_preamble(raw: bytes) -> bytes:
        count = len(raw) - len(raw.rstrip(b"\xFE"))
        if not count:
            return b""
        return raw[-min(count, DLT645Protocol.MAX_PREAMBLE_LENGTH) :]

    @classmethod
    def _candidate_start(cls, raw: bytes, start_idx: int) -> int:
        """返回候选帧前连续前导码的起始位置。"""
        candidate_start = start_idx
        while (
            candidate_start > 0
            and start_idx - candidate_start < cls.MAX_PREAMBLE_LENGTH
            and raw[candidate_start - 1] == 0xFE
        ):
            candidate_start -= 1
        return candidate_start

    @classmethod
    def deserialize_with_remaining(cls, raw: bytes) -> tuple[bytes, Optional[Frame]]:
        """从字节流中解析一帧，并返回帧后的剩余数据。

        无效噪声和损坏帧会被丢弃；不完整帧会从首个有效候选 ``0x68``
        起保留，供下一批数据继续拼接。
        """
        raw = bytes(raw)
        search_from = 0

        while True:
            start_idx = raw.find(bytes([FRAME_START_BYTE]), search_from)
            if start_idx < 0:
                return cls._trailing_preamble(raw), None

            # 固定头部尚未收齐，必须保留候选帧。
            if len(raw) < start_idx + 10:
                return raw[cls._candidate_start(raw, start_idx) :], None

            # 第二个起始符位置不匹配，当前 0x68 只是噪声或地址内容。
            if raw[start_idx + 7] != FRAME_START_BYTE:
                search_from = start_idx + 1
                continue

            data_len = raw[start_idx + 9]
            data_start = start_idx + 10
            checksum_idx = data_start + data_len
            frame_end = checksum_idx + 2
            if len(raw) < frame_end:
                return raw[cls._candidate_start(raw, start_idx) :], None

            expected_checksum = cls.calculate_checksum(raw[start_idx:checksum_idx])
            if raw[checksum_idx] != expected_checksum:
                log.warning(
                    f"Checksum error, skipping frame starting at index {start_idx}"
                )
                search_from = start_idx + 1
                continue
            if raw[checksum_idx + 1] != FRAME_END_BYTE:
                log.warning(
                    f"End flag error, skipping frame starting at index {start_idx}"
                )
                search_from = start_idx + 1
                continue

            preamble_start = cls._candidate_start(raw, start_idx)

            frame = Frame(
                preamble=raw[preamble_start:start_idx],
                start_flag=FRAME_START_BYTE,
                addr=raw[start_idx + 1 : start_idx + 7],
                ctrl_code=raw[start_idx + 8],
                data_len=data_len,
                data=cls.decode_data(raw[data_start:checksum_idx]),
                check_sum=raw[checksum_idx],
                end_flag=FRAME_END_BYTE,
            )
            log.debug(f"frame: {frame}")
            return raw[frame_end:], frame

    @classmethod
    def serialize(cls, frame: Frame) -> bytes:
        """序列化帧；前导 ``0xFE`` 不参与校验和计算。"""
        if frame.start_flag != FRAME_START_BYTE or frame.end_flag != FRAME_END_BYTE:
            raise ValueError(
                f"invalid start or end flag: {frame.start_flag} {frame.end_flag}"
            )
        if len(frame.addr) != 6:
            raise ValueError("地址长度必须为6字节")
        if not 0 <= frame.ctrl_code <= 0xFF:
            raise ValueError("控制码必须在0~255范围内")
        if len(frame.preamble) > cls.MAX_PREAMBLE_LENGTH or any(
            value != 0xFE for value in frame.preamble
        ):
            raise ValueError("前导字节必须是0~4个0xFE")

        encoded_data = cls.encode_data(bytes(frame.data))
        if len(encoded_data) > cls.MAX_DATA_LENGTH:
            raise ValueError("数据域长度不能超过255字节")

        body = bytearray([FRAME_START_BYTE])
        body.extend(frame.addr)
        body.append(FRAME_START_BYTE)
        body.append(frame.ctrl_code)
        body.append(len(encoded_data))
        body.extend(encoded_data)

        checksum = cls.calculate_checksum(bytes(body))
        body.extend((checksum, FRAME_END_BYTE))
        frame.data_len = len(frame.data)
        frame.check_sum = checksum
        return bytes(frame.preamble) + bytes(body)
